src/org/jitsi/impl/neomedia/transform/fec/FlexFec03Receiver.java

```diff
@@ -0,0 +1,341 @@
+                continue;
+            }
+            if (reconstructor.canRecover())
+            {
+                logger.debug("Attemping recovery of missing sequence " +
+                    "number " + reconstructor.missingSequenceNumber);
+                flexFecPacketsToRemove.add(flexFecPacket.getSequenceNumber());
+                RawPacket recovered = reconstructor.recover();
+                if (recovered != null)
+                {
+                    logger.info("Recovered packet " +
+                        recovered.getSequenceNumber());
+                    statistics.numRecoveredPackets++;
+                    saveMedia(recovered);
+                    pkts = ArrayUtils.insert(recovered, pkts, RawPacket.class);
+                }
+                else
+                {
+                    logger.error("Recovery of packet " +
+                        reconstructor.missingSequenceNumber + " failed even" +
+                        " though it should have been possible");
+                    statistics.failedRecoveries++;
+                }
+            }
+        }
+
+        for (Integer flexFecSeqNum : flexFecPacketsToRemove)
+        {
+            fecPackets.remove(flexFecSeqNum);
+        }
+        return pkts;
+    }
+
+    private static class Reconstructor
+    {
+        /**
+         * All available media packets.
+         */
+        private Map<Integer, RawPacket> mediaPackets;
+
+        /**
+         * The FlexFEC packet to be used for recovery.
+         */
+        private FlexFec03Packet fecPacket = null;
+
+        /**
+         * We can only recover a single missing packet, so when we check
+         * for how many are missing, we'll keep track of the first one we find
+         */
+        int missingSequenceNumber = -1;
+
+
+        /**
+         * The number of missing media packets we've detected for the set
+         * FlexFEC packet
+         */
+        int numMissing = -1;
+
+        /**
+         * Initializes a new instance.
+         * @param mediaPackets the currently available media packets. Note that
+         * this is a reference so it will remain up to date as the map is
+         * filled out by the caller.
+         */
+        Reconstructor(Map<Integer, RawPacket> mediaPackets)
+        {
+            this.mediaPackets = mediaPackets;
+        }
+
+        public boolean complete()
+        {
+            return numMissing == 0;
+        }
+
+        public boolean canRecover()
+        {
+            return numMissing == 1;
+        }
+
+        public void setFecPacket(FlexFec03Packet p)
+        {
+            logger.debug("Have " + mediaPackets.size() + " saved media packets");
+            numMissing = 0;
+            logger.debug("Reconstructor checking if recovery is " +
+                "possible: fec packet " + p.getSequenceNumber() +
+                " protects packets:\n" + p.getProtectedSequenceNumbers());
+            this.fecPacket = p;
+            if (p == null)
```

Comment:
If p were null it would have already produced a NPE in the debug logs above. Also missing docs on the method.

Reply:
nice catch :) fixed